experiments/download_ssl4eo.py

```diff
@@ -0,0 +1,738 @@
+    collection = collection.filter(ee.Filter.lt("CLOUDY_PIXEL_PERCENTAGE", cloud_pct))
+    # collection = collection.map(maskS2clouds)
+    return collection
+
+
+def get_collection_s1() -> ee.ImageCollection:
+    collection = ee.ImageCollection("COPERNICUS/S1_GRD")
+    return collection
+
+
+def filter_collection(
+    collection: ee.ImageCollection, coords: List[float], period: Tuple[str, str]
+) -> ee.ImageCollection:
+    filtered = collection
+    if period is not None:
+        filtered = filtered.filterDate(*period)  # filter time
+    filtered = filtered.filterBounds(ee.Geometry.Point(coords))  # filter region
+
+    if filtered.size().getInfo() == 0:
+        raise ee.EEException(
+            f"ImageCollection.filter: No suitable images found in \
+                ({coords[1]:.4f}, {coords[0]:.4f}) between {period[0]} and {period[1]}."
+        )
+    return filtered
+
+
+def filter_collection_s1(
+    collection: ee.ImageCollection, coords: List[float], period: Tuple[str, str]
+) -> ee.ImageCollection:
+    filtered = collection
+    if period is not None:
+        filtered = filtered.filterDate(*period)  # filter time
+    filtered = filtered.filterBounds(ee.Geometry.Point(coords))  # filter region
+
+    filtered = filtered.filter(
+        ee.Filter.listContains("transmitterReceiverPolarisation", "VV")
+    ).filter(ee.Filter.listContains("transmitterReceiverPolarisation", "VH"))
+    filtered = filtered.filter(ee.Filter.eq("instrumentMode", "IW"))
+    # filtered = filtered.filter(ee.Filter.eq('orbitProperties_pass', 'ASCENDING'))
+    # filtered = filtered.filter(ee.Filter.eq('orbitProperties_pass', 'DESCENDING'))
+
+    if filtered.size().getInfo() == 0:
+        raise ee.EEException(
+            f"ImageCollection.filter: No suitable images found in \
+            ({coords[1]:.4f}, {coords[0]:.4f}) between {period[0]} and {period[1]}."
+        )
+    return filtered
+
+
+def center_crop(
+    img: np.ndarray[Any, np.dtype[Any]], out_size: Tuple[int, int]
+) -> np.ndarray[Any, np.dtype[Any]]:
+    image_height, image_width = img.shape[:2]
+    crop_height, crop_width = out_size
+    crop_top = int((image_height - crop_height + 1) * 0.5)
+    crop_left = int((image_width - crop_width + 1) * 0.5)
+    return img[crop_top : crop_top + crop_height, crop_left : crop_left + crop_width]
+
+
+def adjust_coords(
+    coords: List[List[float]], old_size: Tuple[int, int], new_size: Tuple[int, int]
+) -> List[List[float]]:
+    xres = (coords[1][0] - coords[0][0]) / old_size[1]
+    yres = (coords[0][1] - coords[1][1]) / old_size[0]
+    xoff = int((old_size[1] - new_size[1] + 1) * 0.5)
+    yoff = int((old_size[0] - new_size[0] + 1) * 0.5)
+    return [
+        [coords[0][0] + (xoff * xres), coords[0][1] - (yoff * yres)],
+        [
+            coords[0][0] + ((xoff + new_size[1]) * xres),
+            coords[0][1] - ((yoff + new_size[0]) * yres),
+        ],
+    ]
+
+
+def get_properties(image: ee.Image) -> Any:
```

Comment:
Any idea what the return type actually is? Any should be avoided whenever possible since it disables type checking

Reply:
It should be a dict, but if I modify that I got the error: "Returning Any from function declared to return "Dict[str, Any]""

Reply:
Ah, yes, if ee is missing type hints then it won't know what the real type should be. Correct way to handle this is to cast to the proper type, but that's a lot of work and this code isn't going to be installed in the core library so Any is fine for now.